lib/src/v3/types.dart

```diff
@@ -126,13 +126,14 @@ enum PgDataType<Dart extends Object> {
   /// The object ID of this data type.
   final int? oid;
 
-  Codec<Dart?, Uint8List?> get binaryCodec {
-    return _binaryCodecs.putIfAbsent(this, () => _BinaryTypeCodec<Dart>(this))
+  Codec<Dart?, Uint8List?> binaryCodec(Encoding charset) {
+    return _binaryCodecs.putIfAbsent(
+            this, () => _BinaryTypeCodec<Dart>(this, charset))
         as Codec<Dart?, Uint8List?>;
   }
 
-  Codec<Dart?, Uint8List?> get textCodec {
-    return _textCodecs.putIfAbsent(this, () => _TextTypeCodec<Dart>(this))
+  Codec<Dart?, Uint8List?>  textCodec(Encoding charset)  {
+    return _textCodecs.putIfAbsent(this, () => _TextTypeCodec<Dart>(this,charset))
```

Comment:
The `Encoding` needs to be part of the key in the map as well. Otherwise you risk a situation where you first call `PgDataType.varCharArray.binaryCodec(utf8)` and then later `PgDataType.varCharArray.binaryCodec(ascii)` - you'd get the same codec with the same encoding of the first call because the value is already in the map.

Or perhaps we can continue using a single map, but only use it when `charset is Utf8Codec`? The cache is mainly a performance optimization, and since most users are on utf8, we might not have to support it for every encoding.

Reply:
corrected as below:

```dart
Codec<Dart?, Uint8List?> binaryCodec(Encoding charset) {
    if (charset is Utf8Codec) {
      return _binaryCodecs.putIfAbsent(
              this, () => _BinaryTypeCodec<Dart>(this, charset))
          as Codec<Dart?, Uint8List?>;
    } else {
      return _BinaryTypeCodec<Dart>(this, charset);
    }
  }

  Codec<Dart?, Uint8List?> textCodec(Encoding charset) {
    if (charset is Utf8Codec) {
      return _textCodecs.putIfAbsent(
              this, () => _TextTypeCodec<Dart>(this, charset))
          as Codec<Dart?, Uint8List?>;
    } else {
      return _TextTypeCodec<Dart>(this, charset);
    }
  }
```